chore: add VirtualizedList  best practices (#9886)
artsy/eigen: docs/best_practices.md

For the most part you don't have to worry about this.

## Analytics and tracking

In React-native, we use react-tracking as a wrapper for the tracking events we send to Segment. You can read more about the implementation [here](./analytics_and_tracking.md).

## VirtualizedList best practices

Smoothly rendering lists and animations is crucial for a positive user experience. However, performance issues can arise, causing lag, stuttering, and decreased responsiveness. Below are few tips to achieve that in Eigen:

### Never nest ScrollViews.

If you feel like there is no other way, it's probably a better idea to talk to the designer to adjust the approach they're suggested instead. Our `Screen` wrappers from `palette-mobile` expose performant `ScrollView` based components such as `Flatlist` and `Flashlist` that can be used and would save you the nesting.

### Always default to `Flashlist`.

Think of `Flashlist` as `Flatlist` on steroids. It's fast, performant and easy to use.

> **What if you followed all the above steps and you still had performance issues:?**

> Below are some tips to improve the performance further, don't follow them UNLESS you need to, premature optimisation will haunt you and can lead to issues that are not trivial to debug (plus it's arguably not the best use of your time)

Reviewer: ```suggestion
Below are some tips to improve the performance further; don't follow them UNLESS you need to, as premature optimization will haunt you and can lead to issues that are non-trivial to debug (plus it's arguably not the best use of your time).
```

> ### Use `memo` for the rescue. See: https://reactnative.dev/docs/optimizing-flatlist-configuration#use-memo

Reviewer: ```suggestion
### Use `memo` to the rescue. See: https://reactnative.dev/docs/optimizing-flatlist-configuration#use-memo
```

### Use `windowSize` with caution

### Use `LazyFlatlist` in order to define your own lazy loading logic.

<details>
<summary>Code snippet</summary>

Example of a PR implementing it: https://github.com/artsy/eigen/pull/9832

```typescript
const App () => {
  return (
    <LazyFlatlist<NotificationT> keyExtractor={keyExtractor}>
      {(props) => {
        return (
          <FlatlistComponent
            ...
            renderItem={({ item }) => {
              return <ActivityItem notification={item} isVisible={props.hasSeenItem(item)} />
            }}
            onViewableItemsChanged={props.onViewableItemsChanged}
            viewabilityConfig={props.viewabilityConfig}
          />
        )
      }}
    </LazyFlatlist>
  )
}
```

</details>

### Does your component contain animations?

<details>
<summary>If yes, consider moving all the animations to the native thread, how?</summary>

> Example of a component implementing a fade in animation with Moti.

Reviewer: ```suggestion
Example of a potentially problematic component implementing a fade in animation with Moti.
```

### iOS Navigation

- For native iOS code we use the **ARScreenPresenterModule**. Once the new app shell is finished (expected March 2022), we will be using `react-navigation`.

For the most part you don't have to worry about this.

## Analytics and tracking

In React-native, we use react-tracking as a wrapper for the tracking events we send to Segment. You can read more about the implementation [here](./analytics_and_tracking.md).

## VirtualizedList best practices

Smoothly rendering lists and animations is crucial for a positive user experience. However, performance issues can arise, causing lag, stuttering, and decreased responsiveness. Below are few tips to achieve that in Eigen:

### Never nest ScrollViews.

If you feel like there is no other way, it's probably a better idea to talk to the designer to adjust the approach they're suggested instead. Our `Screen` wrappers from `palette-mobile` expose performant `ScrollView` based components such as `Flatlist` and `Flashlist` that can be used and would save you the nesting.

### Always default to `FlashList`.

> Think of `FlashList` as `Flatlist` on steroids. It's fast, performant and easy to use.

Reviewer 2: 🙃

<details>
<summary>
Code snippet
</summary>

```typescript
import { FlashList } from "@shopify/flash-list"

const App = () => {
  return (
   <FlashList
    renderItem={renderItem}
    estimatedItemSize={ESTIMATED_ITEM_SIZE}
    keyExtractor={keyExtractor}
  >
  )
}
```

</details>
```
```

> **What if you followed all the above steps and still have performance issues?**

> Below are some tips to improve the performance further; don't follow them UNLESS you need to, as premature optimization will haunt you and can lead to issues that are non-trivial to debug (plus it's arguably not the best use of your time).

Reviewer 2: 👏

In React-native, we use react-tracking as a wrapper for the tracking events we send to Segment. You can read more about the implementation [here](./analytics_and_tracking.md).

### VirtualizedList best practices

Smoothly rendering lists and animations is crucial for a positive user experience. However, performance issues can arise, causing lag, stuttering, and decreased responsiveness. Below are few tips to achieve that in Eigen:

#### Never nest ScrollViews. If you feel like there is no other way, it's probably a better idea to talk to the designer to adjust the approach they're suggested instead. Our `Screen` wrappers from `palette-mobile` expose performant `ScrollView` based components such as `Flatlist` and `Flashlist` that can be used and would save you the nesting.

> #### Always default to `Flashlist`. Think of it as `Flatlist` on steroids.

Reviewer: > default to `Flashlist`
Link or small example just below?

Author: good idea

> > **What if you followed all the above steps and you still had performance issues:?**

Reviewer: ```suggestion
> **What if you followed all the above steps and still have performance issues?**
```
